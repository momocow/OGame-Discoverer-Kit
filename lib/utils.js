import { view } from 'emago'

export const url = view.getUrl()

export const universeSlug = view.getMeta('ogame-universe')
export const playerId = view.getPlayerId()

export function* range (start, end) {
  if (typeof end === 'undefined') {
    end = start
    start = 0
  }
  for (let i = start; i < end; i++) {
    yield i
  }
}

export function isIngamePage (name) {
  return url.searchParams.get('page') === name
}

function* _progressGen ({ start = 0 , finish = 100 } = {}) {
  let progress = start
  let progObj
  while (progress < finish) {
    const { step, value, message = '', ...props } = (yield progObj) ?? {}
    if (typeof value !== 'undefined') { progress = value }
    if (typeof step !== 'undefined') { progress += step }
    progress = Math.max(Math.min(progress, finish), start)
    progObj = Object.assign({}, props, { progress, message })
  }
  return progObj
}

export function createProgress (...args) {
  const gen = _progressGen(...args)
  gen.next() // consume the first yield
  return gen
}


export const goog = {
  /**
   * Returns a string with at least 64-bits of randomness.
   *
   * Doesn't trust Javascript's random function entirely. Uses a combination of
   * random and current timestamp, and then encodes the string in base-36 to
   * make it shorter.
   *
   * @return {string} A random string, e.g. sn1s7vb4gcic.
   * 
   * @see {@link https://github.com/google/closure-library/blob/555e0138c83ed54d25a3e1cd82a7e789e88335a7/closure/goog/string/string.js#L1177}
   */
  getRandomString () {
    const x = 2147483648
    return Math.floor(Math.random() * x).toString(36) +
          Math.abs(Math.floor(Math.random() * x) ^ Date.now()).toString(36);
  }
}

