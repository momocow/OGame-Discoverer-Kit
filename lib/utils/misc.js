export function* range (arg1, arg2, arg3) {
  let start, end, options
  if (typeof arg2 === 'number') {
    start = arg1
    end = arg2
    options = arg3
  } else {
    start = 0
    end = arg1
    options = arg2
  }
  const { step } = Object.assign({ step: 1 }, options)
  let i = start
  for (; i < end; i += step) { yield i }
  return i
}

function* _progressGen ({ start = 0 , finish = 100 } = {}) {
  let progress = start
  let progObj
  while (progress < finish) {
    const { step, value, message = '', ...props } = (yield progObj) ?? {}
    if (typeof value !== 'undefined') { progress = value }
    if (typeof step !== 'undefined') { progress += step }
    progress = Math.max(Math.min(progress, finish), start)
    progObj = Object.assign({}, props, { progress, message })
  }
  return progObj
}

export function createProgress (...args) {
  const gen = _progressGen(...args)
  gen.next() // consume the first yield
  return gen
}

export class DefaultMap extends Map {
  constructor (factory, ...args) {
    super(...args)
    this.factory = factory
  }

  get (key) {
    if (!super.has(key)) {
      super.set(key, this.factory())
    }
    return super.get(key)
  }
}
