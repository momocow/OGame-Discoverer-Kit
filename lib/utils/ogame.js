import { view } from 'emago'

export const url = view.getUrl()
export const universeSlug = view.getMeta('ogame-universe')
export const playerId = view.getPlayerId()

export function isIngamePage (name) {
  return url.searchParams.get('page') === name
}
