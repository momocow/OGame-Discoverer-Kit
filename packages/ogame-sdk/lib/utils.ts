export function $parseInt (str: string, radix?: number): number {
  const n = parseInt(str, radix)
  if (Number.isNaN(n)) {
    throw new Error('parseInt: NaN')
  }
  return n
}

export function $parseFloat (str: string): number {
  const n = parseFloat(str)
  if (Number.isNaN(n)) {
    throw new Error('parseFloat: NaN')
  }
  return n
}
