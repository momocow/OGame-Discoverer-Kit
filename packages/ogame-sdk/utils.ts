import _ from 'lodash'

export interface NumeralFormatOptions {
  fragment: number
  decimalPoint: string
  delimiter: string
}

/**
 * OGame number formatted with fragmentation for readability.
 */
export class Numeral extends Number {
  public static formatOptions: NumeralFormatOptions = {
    fragment: 3,
    decimalPoint: '.',
    delimiter: ','
  }

  public format (options?: Partial<NumeralFormatOptions>): string {
    const {
      fragment, decimalPoint, delimiter
    } = Object.assign({}, Numeral.formatOptions, options)
    const [integer, fraction = ''] = super.toString().split('.')
    let tokens: string[] = []
    for (
      let idx = 0, len = (integer.length % fragment) || fragment;
      idx + len <= integer.length;
      idx += len, len = fragment
    ) {
      tokens.push(integer.substr(idx, len))
    }
    return tokens.join(delimiter) + (fraction ? decimalPoint + fraction : '')
  }
  
  public toString (): string {
    return this.format()
  }

  public static parse (
    value: string,
    options?: Partial<NumeralFormatOptions>
  ): Numeral {
    const {
      decimalPoint,
      delimiter
    } = Object.assign({}, Numeral.formatOptions, options)
    return new Numeral(value
      .replace(new RegExp(_.escapeRegExp(decimalPoint)), '.')
      .replace(new RegExp(_.escapeRegExp(delimiter), 'g'), '')
    )
  }
}