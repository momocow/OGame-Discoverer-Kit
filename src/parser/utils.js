export function bimap (obj) {
  const inverse = Object.entries(obj)
    .reduce((o, [k, v]) => Object.assign(o, { [v]: k }), {})
  return new Proxy(obj, {
    get (target, prop) {
      return prop in target ? target[prop] : inverse[prop]
    }
  })
}
