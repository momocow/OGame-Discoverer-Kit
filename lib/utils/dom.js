import { EasterEgg } from '../errors'
import { goog } from './misc'

const elementIdSet = new Set()

export function generateElementId (categoryOrOptions) {
  let label = 'ogdk'
  let category = ''
  let maxTry = 10
  if (typeof categoryOrOptions === 'object') {
    label = categoryOrOptions.label ?? label
    category = categoryOrOptions.category ?? category
  } else {
    category = categoryOrOptions
  }

  let nodeId
  let tries = 0
  do {
    nodeId = [ label, category, goog.getRandomString() ]
      .filter(Boolean)
      .join('-')
    tries++
  } while (elementIdSet.has(nodeId) && tries < maxTry)

  if (elementIdSet.has(nodeId)) {
    throw new EasterEgg('Failed to generate node ID.')
  }

  elementIdSet.add(nodeId)
  return nodeId
}
