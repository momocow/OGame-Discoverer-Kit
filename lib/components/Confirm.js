import React from 'react'
import PropTypes from 'prop-types'
import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogTitle from '@material-ui/core/DialogTitle'

import { generateElementId } from '@lib/utils/dom'

export default function Confirm ({
  open,
  onClose,
  onConfirm,
  onCancel,
  title,
  message,
  confirmText,
  cancelText,
  children
}) {
  const [ titleId ] = React.useState(generateElementId)
  const [ descId ] = React.useState(generateElementId)
  return (
    <Dialog
      open={open}
      disableBackdropClick
      onClose={onClose}
      aria-labelledby={titleId}
      aria-describedby={descId}
    >
      <DialogTitle id={titleId}>{title}</DialogTitle>
      <DialogContent>
        <DialogContentText id={descId}>{message}</DialogContentText>
        {children}
      </DialogContent>
      <DialogActions>
        <Button onClick={onConfirm} color="primary" autoFocus>
          {confirmText}
        </Button>
        <Button onClick={onCancel}>
          {cancelText}
        </Button>
      </DialogActions>
    </Dialog>
  )
}

Confirm.propTypes = {
  open: PropTypes.bool.isRequired,
  onClose: PropTypes.func,
  onConfirm: PropTypes.func,
  onCancel: PropTypes.func,
  title: PropTypes.string,
  message: PropTypes.string,
  confirmText: PropTypes.string,
  cancelText: PropTypes.string,
  children: PropTypes.node
}
