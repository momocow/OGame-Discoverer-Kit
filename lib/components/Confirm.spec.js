import React from 'react'
import Confirm from './Confirm'
import enzymeToJSON from 'enzyme-to-json'
import { shallow } from 'enzyme'
import { goog } from '../utils/misc'

jest.mock('../utils/misc')

test('<Confirm />', () => {
  const title = 'Confirm Test'
  const message = 'This is a test for Confirm component.'

  let seed = 0
  goog.getRandomString.mockImplementation(() => String(seed++))

  const closedConfirm = shallow(
    <Confirm open={false} title={title} message={message} />
  )
  expect(enzymeToJSON(closedConfirm)).toMatchSnapshot()

  const openedConfirm = shallow(
    <Confirm open={true} title={title} message={message} />
  )
  expect(enzymeToJSON(openedConfirm)).toMatchSnapshot()

  goog.getRandomString.mockRestore()
})
