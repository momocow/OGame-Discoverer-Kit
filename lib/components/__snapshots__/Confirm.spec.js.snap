// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Confirm /> 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-describedby="ogdk-1"
  aria-labelledby="ogdk-0"
  disableBackdropClick={true}
  open={false}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="ogdk-0"
  >
    Confirm Test
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogContentText))
      id="ogdk-1"
    >
      This is a test for Confirm component.
    </WithStyles(ForwardRef(DialogContentText))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      autoFocus={true}
      color="primary"
    />
    <WithStyles(ForwardRef(Button)) />
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`<Confirm /> 2`] = `
<WithStyles(ForwardRef(Dialog))
  aria-describedby="ogdk-3"
  aria-labelledby="ogdk-2"
  disableBackdropClick={true}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="ogdk-2"
  >
    Confirm Test
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogContentText))
      id="ogdk-3"
    >
      This is a test for Confirm component.
    </WithStyles(ForwardRef(DialogContentText))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      autoFocus={true}
      color="primary"
    />
    <WithStyles(ForwardRef(Button)) />
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
